File: stack/0331_verify_preorder_serialization_of_a_binary_tree/0331_verify_preorder_serialization_of_a_binary_tree.go
package main 

import (
	"fmt"
	"strings"
)


func isValidSerialization1(preorder string) bool {
	nodes, diff := strings.Split(preorder, ","), 1
	for _, node := range nodes {
		diff--
		if diff < 0 {
			return false
		}
		if node != "#" {
			diff += 2
		}
	}
	return diff == 0
}

func main() {
	preorder := "9,3,4,#,#,1,#,#,2,#,6,#,#"
	res := isValidSerialization1(preorder)
	fmt.Println(res)
}

File: stack/0331_verify_preorder_serialization_of_a_binary_tree/0331_verify_preorder_serialization_of_a_binary_tree.py

class Solution(object):
    def isValidSerialization(self, preorder):
        p = preorder.split(',')
        slot = 1
        for node in p:
            if slot == 0:
                return False
            if node == '#':
                slot -= 1
            else:
                slot += 1
        
        return slot == 0


preorder = "9,3,4,#,#,1,#,#,2,#,6,#,#"
res = Solution().isValidSerialization(preorder)
print(res)